src/admin/fill_top_versions.rs

```diff
@@ -0,0 +1,82 @@
+use crate::models::Crate;
+use crate::{db, schema::crates};
+use anyhow::Context;
+use diesel::prelude::*;
+
+#[derive(clap::Parser, Debug)]
+#[clap(
+    name = "fill-top-versions",
+    about = "Iterates over every crate ever uploaded and updates their \
+        `highest_version`, `highest_stable_version` and `newest_version` columns.",
+    after_help = "Warning: this can take a lot of time."
+)]
+pub struct Opts {
+    /// How many crates should be queried and processed at a time.
+    #[clap(long, default_value = "25")]
+    page_size: u32,
+}
+
+pub fn run(opts: Opts) -> anyhow::Result<()> {
+    tracing_subscriber::fmt().init();
+
+    let conn = db::connect_now().context("Failed to connect to the database")?;
+
+    let num_crates: i64 = crates::table
+        .count()
+        .get_result(&conn)
+        .context("Failed to count crates in the database")?;
+
+    let page_size = opts.page_size as i64;
+    let pages = num_crates / page_size + 1;
+
+    for page in 1..=pages {
+        info!("Processing page {} / {}", page, pages);
+
+        let crates: Vec<Crate> = Crate::all()
+            .order(crates::id.asc())
+            .limit(page_size)
+            .offset(page_size * (page - 1))
+            .load(&conn)
+            .context("Failed to load crates")?;
```

Comment:
It's probably easier and more performant here to do offset-pagination (`WHERE id > last_id`), but it shouldn't matter *that* much.

Reply:
yeah, guess that could work too. it's running pretty fast already though, so probably not needed with our current number of crates.